fix(@opentelemetry/exporter-collector): remove fulfilled promises cor…

aabmass opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/master/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/master/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?
Fixes #1774 and fixes the issue in several other places

## Short description of the changes
Move the `_onFinish()` callback code, which pops the resolved promises off the queue, to a separate `then()` where a reference to `promise` is available.

One strange thing here is that the popping code will not run at shutdown with `Promise.all(this._sendingPromises)` because it is not the same promise pushed in the queue. I think this should be ok though?

obecny reviewed (commented):
it looks good, just wondering one thing here, if `onSuccess` or `onError` for any reason will raise an error the `resolve` will never be called and the promise will stay there forever. What if you resolve the promise first and then call the callback so it will never be a problem ?

aabmass commented:
@obecny I don't think I've changed that behavior in this PR, but I am happy to modify.

Would moving the splice to a `.finally()` work as well?

obecny commented:
> @obecny I don't think I've changed that behavior in this PR, but I am happy to modify.
> 
> If onSuccess/onError raise an error, then the promise would reject right? Then, would moving the splice to a `.finally()` work as well?

I know you haven't changed that, but just analyzing all edge cases and this is the last thing that might produce such behaviour with not removing the promise probably very rare but still.

With regards to `finally` I think it will not change anything in whole flow.

aabmass commented:
@obecny I changed it to resolve the promise first and then moved the `onSuccess`/`onError` calls into separate continuations, lmk how that looks.

I also added a test in 4d995cf4db9256a323e5ad90147b38e2b84471e9 for this

vmarchaud reviewed (approved):
lgtm, could you fix the test ? thanks

aabmass commented:
@vmarchaud any idea what the issue is? It only fails on node8, I think it is something node8 specific.

dyladan commented on packages/opentelemetry-exporter-collector-grpc/src/CollectorExporterNodeBase.ts:
```diff
@@ -55,25 +55,17 @@ export abstract class CollectorExporterNodeBase<
+    const promise = new Promise<void>((resolve, reject) => {
+      this._send(this, objects, resolve, reject);
+    })
+      .then(onSuccess)
+      .catch(onError);
```
If the `onSuccess` callback throws, the `onError` callback will be called unexpectedly. Passing them both to `.then` will ensure `onError` is only called if the original promise throws.

```suggestion
      .then(onSuccess, onError);
```

aabmass replied:
Nice catch

dyladan commented on packages/opentelemetry-exporter-collector-proto/src/CollectorExporterNodeBase.ts:
```diff
@@ -36,25 +36,17 @@ export abstract class CollectorExporterNodeBase<
+    const promise = new Promise<void>((resolve, reject) => {
+      this._send(this, objects, resolve, reject);
+    })
+      .then(onSuccess)
+      .catch(onError);
```
```suggestion
      .then(onSuccess, onError);
```

dyladan commented on packages/opentelemetry-exporter-collector/src/platform/browser/CollectorExporterBrowserBase.ts:
```diff
@@ -76,27 +76,20 @@ export abstract class CollectorExporterBrowserBase<
       }
-    });
+    })
+      .then(onSuccess)
+      .catch(onError);
```
```suggestion
      .then(onSuccess, onError);
```

dyladan commented on packages/opentelemetry-exporter-collector/src/platform/node/CollectorExporterNodeBase.ts:
```diff
@@ -68,30 +68,23 @@ export abstract class CollectorExporterNodeBase<
       );
-    });
+    })
+      .then(onSuccess)
+      .catch(onError);
```
```suggestion
      .then(onSuccess, onError);
```

dyladan commented:
Overall LGTM but the tests are still failing

aabmass commented:
Sorry for the slow progress here. I didn't realize `Promise.finally()` wasn't available in node 8. I've replaced those areas with the two parameter `.then()`.

Rauno56 commented on packages/opentelemetry-exporter-collector-grpc/test/CollectorExporterNodeBase.test.ts:
```diff
@@ -0,0 +1,156 @@
+}
+
+// Mocked _send which just saves the callbacks for later
+MockCollectorExporter.prototype['_send'] = function _sendMock(
```
Is square bracket notation used here to bypass eslint's convention not to access `_` prefixed properties?

dyladan replied:
it's actually to access private fields. Something we allow in tests

obecny reviewed (approved):
lgtm
